rr_graph_uxsdcxx.cpp

```diff
@@ -0,0 +1,3174 @@
+			break;
+		default: break; /* Not possible. */
+		}
+	}
+	std::bitset<5> test_astate = astate | std::bitset<5>(0b00000);
+	if(!test_astate.all()) attr_error(test_astate, atok_lookup_t_channel);
+}
+
+void load_x_list(const pugi::xml_node &root, t_x_list *out){
+	if(root.first_child().type() == pugi::node_element)
+		throw std::runtime_error("Unexpected child element in <x_list>.");
+
+	std::bitset<2> astate = 0;
+	for(pugi::xml_attribute attr = root.first_attribute(); attr; attr = attr.next_attribute()){
+		atok_t_x_list in = alex_t_x_list(attr.name());
+		if(astate[(int)in] == 0) astate[(int)in] = 1;
+		else throw std::runtime_error("Duplicate attribute " + std::string(attr.name()) + " in <x_list>.");
+		switch(in){
+		case atok_t_x_list::INDEX:
+			out->index = std::strtoul(attr.value(), NULL, 10);
+			if(errno != 0)
+				throw std::runtime_error("Invalid value `" + std::string(attr.value()) + "` to load a unsigned int into out->index.");
+			break;
+		case atok_t_x_list::INFO:
+			out->info = std::strtol(attr.value(), NULL, 10);
+			if(errno != 0)
+				throw std::runtime_error("Invalid value `" + std::string(attr.value()) + "` to load a int into out->info.");
+			break;
+		default: break; /* Not possible. */
+		}
+	}
+	std::bitset<2> test_astate = astate | std::bitset<2>(0b00);
+	if(!test_astate.all()) attr_error(test_astate, atok_lookup_t_x_list);
+}
+
+void load_y_list(const pugi::xml_node &root, t_y_list *out){
+	if(root.first_child().type() == pugi::node_element)
+		throw std::runtime_error("Unexpected child element in <y_list>.");
+
+	std::bitset<2> astate = 0;
+	for(pugi::xml_attribute attr = root.first_attribute(); attr; attr = attr.next_attribute()){
+		atok_t_y_list in = alex_t_y_list(attr.name());
+		if(astate[(int)in] == 0) astate[(int)in] = 1;
+		else throw std::runtime_error("Duplicate attribute " + std::string(attr.name()) + " in <y_list>.");
+		switch(in){
+		case atok_t_y_list::INDEX:
+			out->index = std::strtoul(attr.value(), NULL, 10);
+			if(errno != 0)
+				throw std::runtime_error("Invalid value `" + std::string(attr.value()) + "` to load a unsigned int into out->index.");
+			break;
+		case atok_t_y_list::INFO:
+			out->info = std::strtol(attr.value(), NULL, 10);
+			if(errno != 0)
+				throw std::runtime_error("Invalid value `" + std::string(attr.value()) + "` to load a int into out->info.");
+			break;
+		default: break; /* Not possible. */
+		}
+	}
+	std::bitset<2> test_astate = astate | std::bitset<2>(0b00);
+	if(!test_astate.all()) attr_error(test_astate, atok_lookup_t_y_list);
+}
+
+void load_timing(const pugi::xml_node &root, t_timing *out){
+	if(root.first_child().type() == pugi::node_element)
+		throw std::runtime_error("Unexpected child element in <timing>.");
+
+	std::bitset<5> astate = 0;
+	for(pugi::xml_attribute attr = root.first_attribute(); attr; attr = attr.next_attribute()){
+		atok_t_timing in = alex_t_timing(attr.name());
+		if(astate[(int)in] == 0) astate[(i
```

Comment:
You do this error pattern quite a bit, probably could be a macro?
```cpp
if(errno != 0)
	throw std::runtime_error("Invalid value `" + std::string(attr.value()) + "` to load a unsigned int into out->id.");
```

Reply:
Good idea, `THROW_IF_ERRNO(attr.value(), "unsigned int", "out->id");`

Reply:
Sounds good.